src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -0,0 +1,316 @@
+        const bound_function_handle = mono_wasm_get_js_handle(bound_fn)!;
+        setI32(function_js_handle, <any>bound_function_handle);
+        return MonoObjectNull;
+    } catch (ex) {
+        return wrap_error(is_exception, ex);
+    }
+}
+
+export function mono_wasm_invoke_bound_function(function_js_handle: JSHandle, data: VoidPtr): InvokeJSResult {
+    const js_fn = mono_wasm_get_jsobj_from_js_handle(function_js_handle);
+    assert(js_fn, `Unknow function for handle ${function_js_handle}`);
+    return js_fn(data);
+}
+
+
+
+function mono_wasm_lookup_function(function_id: MonoString): Function {
+    let res = _invoke_js_function_cache.get(function_id);
+    if (res !== undefined) {
+        return res;
+    }
+
+    const js_function_name = conv_string(function_id);
+    assert(js_function_name, "js_function_name must be string");
+    res = _walk_global_scope_to_find_function(js_function_name);
+    _invoke_js_function_cache.set(function_id, res);
+
+    return res;
+}
+
+function covert_arg_to_js(signature: VoidPtr, data: VoidPtr, position: number): any {
+    //console.log("covert_arg_to_js A " + (<any>signature + (position << 2)));
+    const mono_type = <MonoType><any>getU32(<any>signature + (position << 2));
+    //console.log("covert_arg_to_js B " + mono_type);
+    const covertor = cs_to_js_map.get(mono_type);
+    //console.log("covert_arg_to_js C " + typeof covertor);
+    assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
+    //console.log("covert_arg_to_js D");
+    const r = covertor(<any>data + (position << 2));
+    //console.log("covert_arg_to_js E");
+    return r;
+}
+
+function covert_result_to_cs(signature: VoidPtr, data: VoidPtr, value: any): void {
+    //console.log("covert_result_to_cs A");
+    const mono_type = <MonoType><any>getU32(<any>signature);
+    //console.log("covert_result_to_cs B " + mono_type);
+    if (mono_type === MonoTypeNull) {
+        return;// void
+    }
+    console.log("covert_result_to_cs C");
+    const covertor = js_to_cs_map.get(mono_type);
+    console.log("covert_result_to_cs D");
+    assert(covertor, `Unknow covertor for type ${mono_type} at ret`);
+    console.log("covert_result_to_cs E");
+    covertor(<any>data, value);
+    console.log("covert_result_to_cs F");
+}
+
+function _marshal_int32_to_cs(target: VoidPtr, value: any) {
+    return setI32(target, value);
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function _marshal_void_to_cs(_target: VoidPtr, _value: any) {
+    return 0;
+}
+
+function _marshal_int32_to_js(source: VoidPtr) {
+    //console.log("_marshal_int32_to_js S" + source);
+    const pRef = getU32(source);
+    //console.log("_marshal_int32_to_js P" + pRef);
+    const value = getI32(pRef);
+    //console.log("_marshal_int32_to_js v" + value);
+    return value;
+}
+
+function _marshal_string_to_cs(target: VoidPtr, value: any) {
+    const pError = js_string_to_mono_string(value)!;
```

Comment:
What does this postfix ```!``` do? I can't find it documented on mdn.

Reply:
Claim it's not null. TS only concept.